crates/red_knot_python_semantic/src/types.rs

```diff
@@ -393,6 +400,62 @@ impl<'db> Type<'db> {
         }
     }
 
+    /// Return the `Truthy` subset of this type (intersection with Truthy).
+    #[must_use]
+    pub fn truthy(&self, db: &'db dyn Db) -> Type<'db> {
+        IntersectionBuilder::build_truthy(db, *self)
+    }
+
+    /// Return the `Falsy` subset of this type (intersection with Falsy).
+    #[must_use]
+    pub fn falsy(&self, db: &'db dyn Db) -> Type<'db> {
+        self.falsy_set(db)
+            .unwrap_or_else(|| IntersectionBuilder::build_falsy(db, *self))
+    }
+
+    /// Returns, if it can be expressed, the set of values that are falsy in this type. This is
+    /// defined for some builtin types (e.g. int, str, ...)
+    #[must_use]
+    pub fn falsy_set(&self, db: &'db dyn Db) -> Option<Type<'db>> {
+        match self {
+            // Some builtin types have known falsy values
+            Self::Instance(class) if class.is_known(db, KnownClass::Bool) => {
+                Some(Self::BooleanLiteral(false))
+            }
+            Self::Instance(class) if class.is_known(db, KnownClass::Int) => {
+                Some(Self::IntLiteral(0))
+            }
+            Self::Instance(class) if class.is_known(db, KnownClass::Str) => {
+                Some(Self::StringLiteral(StringLiteralType::new(db, "".into())))
+            }
+            Self::Instance(class) if class.is_known(db, KnownClass::Bytes) => {
+                Some(Self::BytesLiteral(BytesLiteralType::new(db, vec![].into())))
+            }
+            Self::Instance(class) if class.is_known(db, KnownClass::Tuple) => {
+                Some(Self::Tuple(TupleType::new(db, vec![].into())))
+            }
+            Type::LiteralString => KnownClass::Str.to_instance(db).falsy_set(db),
+            Type::IntLiteral(_) => KnownClass::Int.to_instance(db).falsy_set(db),
+            Type::BooleanLiteral(_) => KnownClass::Bool.to_instance(db).falsy_set(db),
+            _ => None,
+        }
```

Comment:
There's a few issues here:
1. An important correctness bug
2. Some missed opportunities where we can give some more precise answers
3. Some opportunities for simplification

The correctness bug is in your treatment of `IntLiteral` and `BooleanLiteral` branches. If a value has `Type::IntLiteral(_)`, it doesn't mean that it's an _arbitrary_ literal integer. We don't currently have a way of expressing such a type in our model (and possibly never will) -- if we did, it would probably be called `LiteralInt`, analogous to our `LiteralString` type. Rather, if a value has `Type::IntLiteral(_)`, it indicates that it is a _specific_ (known) literal integer: `Type::IntLiteral(2)` represents the specific value `2`, and `Type::IntLiteral(3)` represents the specific value `3`. As such, it's not true that `Type::IntLiteral(_).falsey_set(db) == KnownClass::Int.to_instance(db).falsey_set(db)`. If `self` is `Type::IntLiteral(3)`, then the set of possible values that `self` represents has size exactly 1, and the subset of those possible values that are falsey is the empty set (which is `Type::Never`).

The missed opportunity for more precise answers regards arbitrary instances that have `__bool__` methods annotated as returning `Literal[False]`. For these instances (which we haven't implemented the dunder-method lookup for yet in `Type::bool`), we know they are always falsey, so we know that the subset of possible falsey values is exactly equal to the original type. (For an always-falsy type `X`, `X & Falsy == X`; for an always-truthy type `Y`, `Y & Falsy == Never`.)

The opportunity for simplification is that I think it would be more readable (and possibly more efficient) if you used an inner match statement using `ClassType::known` rather than multiple branches using `ClassType::is_known`.

Putting all this together, I would make these changes:

```diff
--- a/crates/red_knot_python_semantic/src/types.rs
+++ b/crates/red_knot_python_semantic/src/types.rs
@@ -419,25 +419,24 @@ impl<'db> Type<'db> {
     pub fn falsy_set(&self, db: &'db dyn Db) -> Option<Type<'db>> {
         match self {
             // Some builtin types have known falsy values
-            Self::Instance(class) if class.is_known(db, KnownClass::Bool) => {
-                Some(Self::BooleanLiteral(false))
-            }
-            Self::Instance(class) if class.is_known(db, KnownClass::Int) => {
-                Some(Self::IntLiteral(0))
-            }
-            Self::Instance(class) if class.is_known(db, KnownClass::Str) => {
-                Some(Self::StringLiteral(StringLiteralType::new(db, "".into())))
-            }
-            Self::Instance(class) if class.is_known(db, KnownClass::Bytes) => {
-                Some(Self::BytesLiteral(BytesLiteralType::new(db, vec![].into())))
-            }
-            Self::Instance(class) if class.is_known(db, KnownClass::Tuple) => {
-                Some(Self::Tuple(TupleType::new(db, vec![].into())))
-            }
-            Type::LiteralString => KnownClass::Str.to_instance(db).falsy_set(db),
-            Type::IntLiteral(_) => KnownClass::Int.to_instance(db).falsy_set(db),
-            Type::BooleanLiteral(_) => KnownClass::Bool.to_instance(db).falsy_set(db),
-            _ => None,
+            Self::Instance(class) => match class.known(db) {
+                Some(KnownClass::Bool) => Some(Self::BooleanLiteral(false)),
+                Some(KnownClass::Int) => Some(Self::IntLiteral(0)),
+                Some(KnownClass::Bytes) => Some(Self::BytesLiteral(BytesLiteralType::empty(db))),
+                Some(KnownClass::Str) => Some(Self::StringLiteral(StringLiteralType::empty(db))),
+                Some(KnownClass::Tuple) => Some(Self::Tuple(TupleType::empty(db))),
+                _ => match self.bool(db) {
+                    Truthiness::AlwaysFalse => Some(*self),
+                    Truthiness::AlwaysTrue => Some(Type::Never),
+                    Truthiness::Ambiguous => None,
+                },
+            },
+            Type::LiteralString => Some(Self::StringLiteral(StringLiteralType::empty(db))),
+            _ => match self.bool(db) {
+                Truthiness::AlwaysFalse => Some(*self),
+                Truthiness::AlwaysTrue => Some(Type::Never),
+                Truthiness::Ambiguous => None,
+            },
         }
     }
 
@@ -1593,18 +1592,36 @@ pub struct StringLiteralType<'db> {
     value: Box<str>,
 }
 
+impl<'db> StringLiteralType<'db> {
+    pub fn empty(db: &'db dyn Db) -> Self {
+        Self::new(db, Box::default())
+    }
+}
+
 #[salsa::interned]
 pub struct BytesLiteralType<'db> {
     #[return_ref]
     value: Box<[u8]>,
 }
 
+impl<'db> BytesLiteralType<'db> {
+    pub fn empty(db: &'db dyn Db) -> Self {
+        Self::new(db, Box::default())
+    }
+}
+
 #[salsa::interned]
 pub struct TupleType<'db> {
     #[return_ref]
     elements: Box<[Type<'db>]>,
 }
 
+impl<'db> TupleType<'db> {
+    pub fn empty(db: &'db dyn Db) -> Self {
+        Self::new(db, Box::default())
+    }
+}
+
```

I think similar comments would apply to your `truthy_set` method below.

Reply:
No problem for the syntax improvements, and for the logic bug, good catch! I added some comments of my understanding of the case for people that will read the code.